### PR TITLE
VIDEO-13386 Performance enhancements

### PR DESCRIPTION
**Pull request by manjeshbhargav**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [ ] I acknowledge that all my contributions will be made under the project's license.

### Description

[Changes](https://github.com/twilio/twilio-video-processors.js/pull/77/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed)

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [x] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [ ] Included screenshot as PR comment (if needed)
* [x] Ready for review

### CODE REVIEW

File: lib/processors/background/BackgroundProcessor.ts [hunk 1]
```diff
@@ -242,42 +256,70 @@ export abstract class BackgroundProcessor extends Processor {
...
+      if (!this._webgl2Pipeline) {
+        this._createWebGL2Pipeline(
+          inputFrameBuffer as HTMLVideoElement,
+          captureWidth,
+          captureHeight,
+          inferenceWidth,
+          inferenceHeight
+        );
+      }
+      this._webgl2Pipeline?.sampleInputFrame();
```
Comment (charliesantos) [inline lib/processors/background/BackgroundProcessor.ts:269]: It seems we're creating _webgl2Pipeline only once. How are we updating the widths and heights params?

File: lib/processors/background/BackgroundProcessor.ts [hunk 2]
```diff
@@ -295,45 +337,34 @@ export abstract class BackgroundProcessor extends Processor {
 
   protected abstract _getWebGL2PipelineType(): WebGL2PipelineType;
 
-  protected abstract _setBackground(inputFrame: OffscreenCanvas | HTMLCanvasElement | HTMLVideoElement): void;
```
Comment (charliesantos) [inline lib/processors/background/BackgroundProcessor.ts:298]: HTMLVideoElement is one of the possible inputs right? Is this a breaking change?

File: lib/processors/webgl2/pipelines/fastBilateralFilterStage.ts [hunk 3]
```diff
@@ -18,6 +18,11 @@ export function buildJointBilateralFilterStage(
   outputTexture: WebGLTexture,
   canvas: HTMLCanvasElement
 ) {
+  // NOTE(mmalavalli): This is a faster approximation of the joint bilateral filter.
+  // It improves the complexity from O(w x h x r^2) to O(w x h x r), where:
+  // w => width of the output video frame
+  // h => height of the output video frame
+  // r => radius of the joint bilateral filter kernel
```
Comment (charliesantos) [inline lib/processors/webgl2/pipelines/fastBilateralFilterStage.ts:25]: Nice. Is this something you wrote or do you have a reference somewhere that provides more details about this algorithm?

**Review by charliesantos — CHANGES_REQUESTED** [review]

File: lib/processors/background/BackgroundProcessor.ts [hunk 4]
```diff
@@ -135,21 +133,28 @@ export abstract class BackgroundProcessor extends Processor {
       assetsPath += '/';
     }
 
-    this.maskBlurRadius = options.maskBlurRadius!;
     this._assetsPath = assetsPath;
-    this._debounce = typeof options.debounce === 'boolean' ? options.debounce : this._debounce;
-    this._debounceCount = this._debounce ? this._debounceCount : 1;
+    this._debounce = typeof options.debounce === 'boolean' ? options.debounce : false;
```
Comment (manjeshbhargav) [inline lib/processors/background/BackgroundProcessor.ts:137]: Set debounce default to true, but emphasize setting it to false in the docs and example code.

File: lib/processors/background/BackgroundProcessor.ts (hunk 1 above)
Comment (manjeshbhargav) [reply lib/processors/background/BackgroundProcessor.ts:269]: For example, when we are setting a new background, we clean up the pipeline, which triggers this code to recreate it with the new height and width in the next `processFrame()` call.

File: lib/processors/background/BackgroundProcessor.ts (hunk 2 above)
Comment (manjeshbhargav) [reply lib/processors/background/BackgroundProcessor.ts:298]: This function is for internal use only. `processFrame()` still accepts an HTMLVideoElement.

**Review by charliesantos — APPROVED** [review]

**Comment by charliesantos** [comment]
+1. Just need to fix unit tests

**Pull request merged**